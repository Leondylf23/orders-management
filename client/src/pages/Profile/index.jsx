import { useDispatch, connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { useIntl, FormattedMessage } from 'react-intl';
import { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import Avatar from '@mui/material/Avatar';

import { showPopup } from '@containers/App/actions';
import { selectUserData } from '@containers/Client/selectors';
import { decryptDataAES, encryptDataAES } from '@utils/allUtils';
import { setUserData } from '@containers/Client/actions';
import ModeEditOutlineIcon from '@mui/icons-material/ModeEditOutline';
import { Link } from 'react-router-dom';

import { selectProfileData } from './selectors';
import { getProfileData, saveProfileData } from './actions';

import classes from './style.module.scss';

const ProfilePage = ({ profileData, userDataSelect }) => {
  console.log(profileData);
  const dispatch = useDispatch();

  const intl = useIntl();

  const [userData, setUserDataInternal] = useState();
  const [profileImg, setProfileImg] = useState(null);

  const setNewProfileImage = (e) => {
    const file = e.target.files[0];
    setProfileImg(file);
  };

  const saveGeneralData = () => {
    if (userData?.fullname.length < 3 || userData?.length > 255) {
      dispatch(
        showPopup(
          intl.formatMessage({ id: 'profile_title' }),
          intl.formatMessage({ id: 'register_fullname_validation' })
        )
      );
      return;
    }
    if (userData?.dob === '') {
      dispatch(
        showPopup(intl.formatMessage({ id: 'profile_title' }), intl.formatMessage({ id: 'register_dob_validation' }))
      );
      return;
    }
    if (new Date().getTime - new Date(userData?.dob).getTime() < 15 * 365 * 24 * 3600000) {
      dispatch(
        showPopup(
          intl.formatMessage({ id: 'profile_title' }),
          intl.formatMessage({ id: 'register_dob_age_validation' })
        )
      );
      return;
    }

    const form = new FormData();
    console.log(userData);

    form.append('fullname', userData?.fullname);
    form.append('dob', userData?.dob);
    form.append('location', userData?.location);

    if (profileImg) form.append('imageData', profileImg);

    dispatch(
      saveProfileData(form, (imageUrl) => {
        if (imageUrl) {
          const user = JSON.parse(decryptDataAES(userDataSelect));
          user.profileImage = imageUrl;

          const updatedUser = encryptDataAES(JSON.stringify(user));
          dispatch(setUserData(updatedUser));
        }

        setProfileImg(null);
        dispatch(
          showPopup(
            intl.formatMessage({ id: 'profile_title' }),
            intl.formatMessage({ id: 'profile_generic_save_success' })
          )
        );
        dispatch(getProfileData());
      })
    );
  };

  useEffect(() => {
    dispatch(getProfileData());
  }, []);
  useEffect(() => {
    setUserDataInternal(profileData);
  }, [profileData]);

  return (
    <div className={classes.mainContainer}>
      <h1 className={classes.title}>
        <FormattedMessage id="profile_title" />
      </h1>
      <div className={classes.contentContainer}>
        <div className={classes.leftSide}>
          <Avatar
            className={classes.profileImage}
            src={profileImg ? URL.createObjectURL(profileImg) : profileData?.profileImage}
            alt="Load image failed!"
          />
          {profileImg ? (
            <button type="button" className={classes.button} data-type="red" onClick={() => setProfileImg(null)}>
              <FormattedMessage id="profile_delete_img" />
            </button>
          ) : (
            <>
              <label htmlFor="profileImageFile" className={classes.button}>
                <FormattedMessage id="profile_chg_img" />
              </label>
              <input type="file" accept="image/*" hidden id="profileImageFile" onChange={setNewProfileImage} />
            </>
          )}
          <div className={classes.accountInfoContainer}>
            <p>
              {profileData?.role === 'customer'
                ? intl.formatMessage({ id: 'profile_customer' })
                : intl.formatMessage({ id: 'profile_business' })}
            </p>
            <p>{profileData?.createdAt}</p>
          </div>
        </div>
        <div className={classes.rigthSide}>
          <h3 className={classes.containerTitle}>
            <FormattedMessage id="profile_general" />
          </h3>
          <label htmlFor="email" className={classes.label}>
            <FormattedMessage id="profile_email" />
          </label>
          <input type="email" id="email" disabled className={classes.input} value={userData?.email} />
          <label htmlFor="fullname" className={classes.label}>
            <FormattedMessage id="profile_fullname" />
          </label>
          <input
            type="text"
            id="fullname"
            className={classes.input}
            value={userData?.fullname}
            onChange={(e) => setUserDataInternal((prevVal) => ({ ...prevVal, fullname: e.target.value }))}
          />
          <label htmlFor="dob" className={classes.label}>
            <FormattedMessage id="profile_dob" />
          </label>
          <input
            type="date"
            id="dob"
            className={classes.input}
            value={userData?.dob}
            onChange={(e) => setUserDataInternal((prevVal) => ({ ...prevVal, dob: e.target.value }))}
          />
          <label htmlFor="dob" className={classes.label}>
            <FormattedMessage id="profile_location" />
          </label>
          <input
            type="text"
            id="location"
            className={classes.input}
            value={userData?.location}
            onChange={(e) => setUserDataInternal((prevVal) => ({ ...prevVal, location: e.target.value }))}
          />
          <label htmlFor="dob" className={classes.label}>
            <FormattedMessage id="profile_passwords" />
          </label>
          <div className={classes.password}>
            <input
              type="password"
              id="dob"
              className={classes.input}
              value="*******"
              onChange={(e) => setUserDataInternal((prevVal) => ({ ...prevVal, dob: e.target.value }))}
              disabled
            />
            <Link to="change-password">
              <ModeEditOutlineIcon />
            </Link>
          </div>
          <div className={classes.buttonConatainer}>
            <button type="button" className={classes.button} onClick={saveGeneralData}>
              <FormattedMessage id="profile_save" />
            </button>
          </div>
        </div>
      </div>
    </div>
  );
};

ProfilePage.propTypes = {
  profileData: PropTypes.object,
  userDataSelect: PropTypes.object,
};

const mapStateToProps = createStructuredSelector({
  profileData: selectProfileData,
  userDataSelect: selectUserData,
});

export default connect(mapStateToProps)(ProfilePage);
