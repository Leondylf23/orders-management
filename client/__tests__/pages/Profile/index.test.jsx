import { fireEvent, render } from '@testing-library/react';
import { Provider } from 'react-redux';
import Profile from '@pages/Profile';

import store from '@store';
import Language from '@containers/Language';
import { MemoryRouter } from 'react-router-dom';

const ParentComponent = (children) => (
  <Provider store={store}>
    <Language>
      <MemoryRouter>{children}</MemoryRouter>
    </Language>
  </Provider>
);

describe('Profile Page', () => {
  beforeEach(() => {});

  test('Correct Render', () => {
    const profile = render(ParentComponent(<Profile />));
    expect(profile.getByTestId('profile-container')).toBeInTheDocument();
    expect(profile.getByTestId('profile-container')).toHaveClass('mainContainer');

    expect(profile).toMatchSnapshot();
  });

  test('Button Clicked', () => {
    const { queryByTestId } = render(ParentComponent(<Profile />));
    const buttonSubmit = queryByTestId('profile-button-submit');
    expect(buttonSubmit).toBeInTheDocument();
    fireEvent.click(buttonSubmit);
  });
});
