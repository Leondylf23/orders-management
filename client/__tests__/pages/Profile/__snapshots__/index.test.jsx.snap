// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Best Seller Card Correct Render 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="mainContainer"
        data-testid="profile-container"
      >
        <h1
          class="title"
        >
          Profile Anda
        </h1>
        <div
          class="contentContainer"
        >
          <div
            class="leftSide"
          >
            <div
              class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault profileImage css-2s90m6-MuiAvatar-root"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
                data-testid="PersonIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
                />
              </svg>
            </div>
            <label
              class="button"
              for="profileImageFile"
            >
              Ganti Gambar
            </label>
            <input
              accept="image/*"
              hidden=""
              id="profileImageFile"
              type="file"
            />
            <div
              class="accountInfoContainer"
            >
              <p>
                Bisnis
              </p>
              <p />
            </div>
          </div>
          <div
            class="rigthSide"
          >
            <h3
              class="containerTitle"
            >
              General
            </h3>
            <label
              class="label"
              for="email"
            >
              Email
            </label>
            <input
              class="input"
              disabled=""
              id="email"
              type="email"
              value=""
            />
            <label
              class="label"
              for="fullname"
            >
              Nama Lengkap
            </label>
            <input
              class="input"
              id="fullname"
              type="text"
              value=""
            />
            <label
              class="label"
              for="dob"
            >
              Tanggal Lahir
            </label>
            <input
              class="input"
              id="dob"
              type="date"
              value=""
            />
            <label
              class="label"
              for="dob"
            >
              Kata Sandi
            </label>
            <div
              class="password"
            >
              <input
                class="input"
                disabled=""
                id="dob"
                type="password"
                value="*******"
              />
              <a
                href="/change-password"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="ModeEditOutlineIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 5.63l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83c.39-.39.39-1.02 0-1.41z"
                  />
                </svg>
              </a>
            </div>
            <div
              class="buttonConatainer"
            >
              <button
                class="button"
                type="button"
              >
                Simpan
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="mainContainer"
      data-testid="profile-container"
    >
      <h1
        class="title"
      >
        Profile Anda
      </h1>
      <div
        class="contentContainer"
      >
        <div
          class="leftSide"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault profileImage css-2s90m6-MuiAvatar-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
              data-testid="PersonIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
              />
            </svg>
          </div>
          <label
            class="button"
            for="profileImageFile"
          >
            Ganti Gambar
          </label>
          <input
            accept="image/*"
            hidden=""
            id="profileImageFile"
            type="file"
          />
          <div
            class="accountInfoContainer"
          >
            <p>
              Bisnis
            </p>
            <p />
          </div>
        </div>
        <div
          class="rigthSide"
        >
          <h3
            class="containerTitle"
          >
            General
          </h3>
          <label
            class="label"
            for="email"
          >
            Email
          </label>
          <input
            class="input"
            disabled=""
            id="email"
            type="email"
            value=""
          />
          <label
            class="label"
            for="fullname"
          >
            Nama Lengkap
          </label>
          <input
            class="input"
            id="fullname"
            type="text"
            value=""
          />
          <label
            class="label"
            for="dob"
          >
            Tanggal Lahir
          </label>
          <input
            class="input"
            id="dob"
            type="date"
            value=""
          />
          <label
            class="label"
            for="dob"
          >
            Kata Sandi
          </label>
          <div
            class="password"
          >
            <input
              class="input"
              disabled=""
              id="dob"
              type="password"
              value="*******"
            />
            <a
              href="/change-password"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="ModeEditOutlineIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75zM20.71 5.63l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83c.39-.39.39-1.02 0-1.41z"
                />
              </svg>
            </a>
          </div>
          <div
            class="buttonConatainer"
          >
            <button
              class="button"
              type="button"
            >
              Simpan
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
